tensorflow/lite/experimental/shlo/ops/dot_general.cc

```diff
@@ -0,0 +1,441 @@
+      op.lhs_index[lhs_contracting_dimensions_data[i]]++;
+      op.rhs_index[rhs_contracting_dimensions_data[i]]++;
+      if (op.lhs_index[lhs_contracting_dimensions_data[i]] <
+          lhs.shape().Dim(lhs_contracting_dimensions_data[i]))
+        return true;
+      if (i == 0) return false;
+      op.lhs_index[lhs_contracting_dimensions_data[i]] = 0;
+      op.rhs_index[rhs_contracting_dimensions_data[i]] = 0;
+    }
+    return true;
+  };
+  // pre compute helper for lhs and rhs indices
+  DimensionSize lhs_dim_accumulator = 1, rhs_dim_accumulator = 1;
+  for (size_t i = 0; i < lhs_rank; ++i) {
+    lhs_dim_accumulator *= lhs.shape().Dim(i);
+    op.lhs_index_helper[i] = lhs_size / lhs_dim_accumulator;
+  }
+  for (size_t i = 0; i < rhs_rank; ++i) {
+    rhs_dim_accumulator *= rhs.shape().Dim(i);
+    op.rhs_index_helper[i] = rhs_size / rhs_dim_accumulator;
+  }
+
+  StorageT output_element(0);
+  DimensionSize lhs_element_index = 0, rhs_element_index = 0;
+  for (DimensionSize k = 0; k < output_size; ++k, ++output_data) {
+    GenerateIndices(k);
+    std::fill(op.lhs_index.begin(), op.lhs_index.end(), 0);
+    std::fill(op.rhs_index.begin(), op.rhs_index.end(), 0);
+    size_t result_dim = 0;
+    for (size_t i = 0; i < lhsb_size; ++i, ++result_dim) {
+      op.lhs_index[lhs_batching_dimensions_data[i]] =
+          op.output_index[result_dim];
+      op.rhs_index[rhs_batching_dimensions_data[i]] =
+          op.output_index[result_dim];
+    }
+    for (size_t i = 0; i < op.lhs_result_dims.size(); ++i, ++result_dim) {
+      op.lhs_index[op.lhs_result_dims[i]] = op.output_index[result_dim];
+    }
+    for (size_t i = 0; i < op.rhs_result_dims.size(); ++i, ++result_dim) {
+      op.rhs_index[op.rhs_result_dims[i]] = op.output_index[result_dim];
+    }
+    output_element = 0;
+    while (true) {
+      lhs_element_index = 0;
+      rhs_element_index = 0;
+      for (size_t i = 0; i < lhs_rank; ++i) {
+        lhs_element_index += op.lhs_index[i] * op.lhs_index_helper[i];
+      }
+      for (size_t i = 0; i < rhs_rank; ++i) {
+        rhs_element_index += op.rhs_index[i] * op.rhs_index_helper[i];
+      }
+      output_element +=
+          lhs_data[lhs_element_index] * rhs_data[rhs_element_index];
+      if (!IncrementIndices()) {
+        break;
+      }
+    }
+    *output_data = output_element;
+  }
+  return absl::OkStatus();
+}
+
+DotGeneralOp Create(DotGeneralOp::Attributes attributes) {
+  return {.attributes = attributes};
+}
+
+absl::Status Prepare(DotGeneralOp& op, const Tensor& lhs, const Tensor& rhs,
+                     Tensor& output) {
+  if (absl::Status status =
+          CheckParameters(lhs, rhs, op.attributes.lhs_batching_dimensions,
+                          op.attributes.rhs_batching_dimensions,
+                          op.attributes.lhs_contracting_dimensions,
+                          op.attributes.rhs_contracting_dimensions, output);
+      !status.ok()) {
+    return status;
+  }
```

Comment:
Use `SHLO_REF_RETURN_ON_ERROR`.

https://github.com/tensorflow/tensorflow/blob/b19a54c0d9739f9c548222d625c1413628b3a43f/tensorflow/lite/experimental/shlo/ops/util.h#L26-L29

```suggestion
  SHLO_REF_RETURN_ON_ERROR(CheckParameters(
                          lhs, rhs, op.attributes.lhs_batching_dimensions,
                          op.attributes.rhs_batching_dimensions,
                          op.attributes.lhs_contracting_dimensions,
                          op.attributes.rhs_contracting_dimensions, output));
```

Reply:
used SHLO_REF_RETURN_ON_ERROR.